File: prototypes/proof_of_concept.py
# Full proof of concept script
# Imports
import cv2
import numpy as np
import json
from kalman_filter import MaskKalmanFilter
import argparse
import torch
from model import UNet
import torch.nn as nn
import threading
from queue import Queue

# --- Custom Exceptions ---
class VideoProcessingError(Exception):
    """Base exception for errors in this script."""
    pass

class VideoCaptureError(VideoProcessingError):
    """Raised when the video file cannot be opened."""
    pass

class FrameProcessingError(VideoProcessingError):
    """Raised for errors during the frame processing loop."""
    pass

class MaskVerificationError(FrameProcessingError):
    """Raised when a generated mask fails the verification step."""
    pass

# --- Threading Functions ---
def frame_reader(video_path: str, frame_queue: Queue, stop_event: threading.Event):
    """
    Reads frames from a video file and puts them into a queue.

    Args:
        video_path: Path to the input video file.
        frame_queue: The queue to store frames.
        stop_event: An event to signal when to stop reading frames.
    """
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        raise VideoCaptureError(f"Cannot open video file: {video_path}")

    while not stop_event.is_set():
        ret, frame = cap.read()
        if not ret:
            break
        frame_queue.put(frame)

    cap.release()
    frame_queue.put(None)  # Sentinel to indicate the end of the stream

def video_processing(video_path: str, output_id: str, model: nn.Module, config: dict) -> None:
    """
    Processes a video to detect and visualize glare, frame by frame, using multithreading.

    Args:
        video_path: Path to the input video file.
        output_id: A unique identifier used for naming the output files.
        model: The pre-trained PyTorch model to use for mask generation.
        config: A dictionary containing configuration parameters.
    """
    # Load configuration
    smoothing_ksize = tuple(config['smoothing']['kernel_size'])
    smoothing_sigma = config['smoothing']['sigma']
    device = config['train']['device']

    model.to(device)
    model.eval()

    # Get video properties for output
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        raise VideoCaptureError(f"Cannot open video file: {video_path}")
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    fps = cap.get(cv2.CAP_PROP_FPS)
    cap.release()

    # Initialize video writers
    mask_writer = create_video_writer(f"{output_id}_mask.mp4", fps, width, height, is_color=False)
    final_writer = create_video_writer(f"{output_id}_final.mp4", fps, width, height, is_color=True)

    # Initialize Kalman Filter
    kf = MaskKalmanFilter(
        shape=(height, width),
        q_val=config['kalman_filter']['Q'],
        r_val=config['kalman_filter']['R']
    )

    # Frame queue and threading events
    frame_queue = Queue(maxsize=int(fps))
    stop_event = threading.Event()

    # Start the frame reader thread
    reader_thread = threading.Thread(target=frame_reader, args=(video_path, frame_queue, stop_event))
    reader_thread.start()

    frame_count = 0
    try:
        while True:
            frame = frame_queue.get()
            if frame is None:  # End of stream
                break

            frame_count += 1
            gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY).astype(np.float32) / 255.0

            mask = generate_mask(model, gray_frame, device)
            mask = smooth_output(mask, smoothing_ksize, smoothing_sigma)
            filtered_mask = kf.update(mask)

            try:
                verify_feedback(filtered_mask)
            except MaskVerificationError as e:
                print(f"Warning: Frame {frame_count} failed verification: {e}")
                continue

            mask_output = (np.clip(filtered_mask, 0.0, 1.0) * 255).astype(np.uint8)
            final_output = finalize_output(frame, filtered_mask)

            mask_writer.write(mask_output)
            final_writer.write(final_output)

    finally:
        # Clean up
        stop_event.set()
        reader_thread.join()
        mask_writer.release()
        final_writer.release()

    if frame_count == 0:
        raise FrameProcessingError("No frames were read from the video.")

    print(f"Processing complete. Videos saved as {output_id}_mask.mp4 and {output_id}_final.mp4")


def generate_mask(model: nn.Module, frame: np.ndarray, device: str) -> np.ndarray:
    """
    Generates a glare mask for a single frame using the U-Net model.

    Args:
        model: The pre-trained PyTorch model.
        frame: The input frame (H, W), normalized to [0, 1].
        device: The device to run the model on ('cuda' or 'cpu').

    Returns:
        The generated mask as a numpy array (H, W).
    """
    # Add batch and channel dimensions for the model
    inp = torch.from_numpy(frame).unsqueeze(0).unsqueeze(0).to(device)
    with torch.no_grad():
        # Squeeze batch and channel dimensions from the output
        out = model(inp).squeeze(0).squeeze(0).cpu().numpy()
    return out


def smooth_output(mask: np.ndarray, ksize: tuple[int, int], sigma: float) -> np.ndarray:
    """
    Applies a Gaussian blur to smooth the mask.

    Args:
        mask: The input mask (H, W).
        ksize: The kernel size for the Gaussian blur.
        sigma: The sigma value for the Gaussian blur.

    Returns:
        The smoothed mask.
    """
    return cv2.GaussianBlur(mask, ksize, sigma)


def verify_feedback(mask: np.ndarray, min_threshold: float = 0.01, max_threshold: float = 0.5):
    """
    Verifies the generated mask based on simple heuristics.

    Checks if the proportion of the activated mask area is within a reasonable range.

    Args:
        mask: The mask to verify.
        min_threshold: The minimum acceptable proportion of the mask being active.
        max_threshold: The maximum acceptable proportion of the mask being active.

    Raises:
        MaskVerificationError: If the mask's active area is outside the defined bounds.
    """
    mask_proportion = np.mean(mask)
    if not (min_threshold < mask_proportion < max_threshold):
        raise MaskVerificationError(
            f"Mask covers {mask_proportion:.2%} of the frame, "
            f"which is outside the acceptable range of ({min_threshold:.2%}, {max_threshold:.2%})."
        )


def finalize_output(frame: np.ndarray, mask: np.ndarray, color: tuple[int, int, int] = (0, 0, 255), alpha: float = 0.4) -> np.ndarray:
    """
    Overlays the mask onto the original frame as a colored highlight.

    Args:
        frame: The original BGR video frame.
        mask: The glare mask (H, W), values in [0, 1].
        color: The BGR color of the overlay.
        alpha: The transparency of the overlay.

    Returns:
        The frame with the mask overlaid.
    """
    overlay = frame.copy()
    colored_mask = (np.stack([mask] * 3, axis=-1) * color).astype(np.uint8)
    cv2.addWeighted(colored_mask, alpha, overlay, 1 - alpha, 0, overlay)
    return overlay


def create_video_writer(filename: str, fps: float, width: int, height: int, is_color: bool) -> cv2.VideoWriter:
    """
    Creates and returns a cv2.VideoWriter object.

    Args:
        filename: The name of the output video file.
        fps: Frames per second for the output video.
        width: Width of the video frames.
        height: Height of the video frames.
        is_color: Flag indicating if the video is color.

    Returns:
        An initialized cv2.VideoWriter object.

    Raises:
        IOError: If the VideoWriter cannot be opened.
    """
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    writer = cv2.VideoWriter(filename, fourcc, fps, (width, height), isColor=is_color)
    if not writer.isOpened():
        raise IOError(f"Failed to open VideoWriter for {filename}")
    return writer


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Process a video file to detect and highlight glare.",
        formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument("video_path", type=str, help="Path to the input video file.")
    parser.add_argument(
        "-o", "--output_id", type=str,
        help="A unique ID for the output files. Defaults to the video's filename."
    )
    parser.add_argument("--config", type=str, default="config.json", help="Path to the config file.")
    args = parser.parse_args()

    video_path = args.video_path
    output_id = args.output_id or video_path.split('/')[-1].split('.')[0]

    try:
        with open(args.config, 'r') as f:
            config = json.load(f)
        
        model = UNet(in_channels=1, out_channels=1)
        model.load_state_dict(torch.load("kvald_unet.pth", map_location=config['train']['device']))

        video_processing(video_path, output_id, model, config)
    except (VideoProcessingError, FileNotFoundError) as e:
        print(f"An error occurred: {e}")

File: src/kvald/proof_of_concept.cpp
#include "kvald/proof_of_concept.h"
#include <iostream>
#include <iomanip>

namespace kvald
{

// --- Threading Functions ---
void frame_reader(
    const std::string& video_path,
    std::queue<cv::Mat>& frame_queue,
    std::mutex& queue_mutex,
    std::condition_variable& queue_cond,
    std::atomic<bool>& stop_reading
)
{
    cv::VideoCapture cap(video_path);
    if (!cap.isOpened())
    {
        stop_reading = true;
        std::lock_guard<std::mutex> lock(queue_mutex);
        frame_queue.push(cv::Mat()); // Sentinel for error/end
        queue_cond.notify_one();
        throw VideoCaptureError("Cannot open video file: " + video_path);
    }

    while (!stop_reading)
    {
        cv::Mat frame;
        if (!cap.read(frame))
        {
            break; // End of video
        }

        std::lock_guard<std::mutex> lock(queue_mutex);
        frame_queue.push(frame);
        queue_cond.notify_one();
    }

    cap.release();
    std::lock_guard<std::mutex> lock(queue_mutex);
    frame_queue.push(cv::Mat()); // Sentinel to indicate the end of the stream
    queue_cond.notify_one();
}

// --- Core Processing Functions ---
cv::Mat generate_mask(
    kvald::UNet& model,
    const cv::Mat& frame,
    torch::Device device
)
{
    // Convert cv::Mat to torch::Tensor
    // OpenCV uses HWC, PyTorch expects NCHW
    torch::Tensor inp = torch::from_blob(
        frame.data,
        {1, frame.rows, frame.cols},
        torch::TensorOptions().dtype(torch::kFloat32)
    );
    inp = inp.to(device);

    torch::Tensor out;
    {
        torch::NoGradGuard no_grad;
        out = model->forward(inp.unsqueeze(0)); // Add batch dimension
    }

    // Convert torch::Tensor back to cv::Mat
    out = out.squeeze().cpu(); // Remove batch and channel dimensions, move to CPU
    cv::Mat mask(out.sizes()[0], out.sizes()[1], CV_32F, out.data_ptr<float>());
    return mask.clone(); // Clone to ensure data ownership
}

cv::Mat smooth_output(
    const cv::Mat& mask,
    const std::vector<int>& ksize,
    double sigma
)
{
    cv::Mat smoothed_mask;
    cv::GaussianBlur(mask, smoothed_mask, cv::Size(ksize[0], ksize[1]), sigma);
    return smoothed_mask;
}

void verify_feedback(
    const cv::Mat& mask,
    float min_threshold,
    float max_threshold
)
{
    double mask_proportion = cv::mean(mask)[0];
    if (!(mask_proportion > min_threshold && mask_proportion < max_threshold))
    {
        std::stringstream ss;
        ss << "Mask covers " << std::fixed << std::setprecision(2) << mask_proportion * 100 << "% of the frame, "
           << "which is outside the acceptable range of (" << min_threshold * 100 << "%, " << max_threshold * 100 << "%).";
        throw MaskVerificationError(ss.str());
    }
}

cv::Mat finalize_output(
    const cv::Mat& frame,
    const cv::Mat& mask,
    const cv::Scalar& color,
    double alpha
)
{
    cv::Mat overlay = frame.clone();
    cv::Mat colored_mask;
    cv::cvtColor(mask, colored_mask, cv::COLOR_GRAY2BGR);
    colored_mask.convertTo(colored_mask, CV_8UC3, 255.0);
    colored_mask = colored_mask.mul(color, 1.0/255.0);

    cv::addWeighted(colored_mask, alpha, overlay, 1 - alpha, 0, overlay);
    return overlay;
}

cv::VideoWriter create_video_writer(
    const std::string& filename,
    double fps,
    int width,
    int height,
    bool is_color
)
{
    int fourcc = cv::VideoWriter::fourcc('m', 'p', '4', 'v');
    cv::VideoWriter writer(filename, fourcc, fps, cv::Size(width, height), is_color);
    if (!writer.isOpened())
    {
        throw std::ios_base::failure("Failed to open VideoWriter for " + filename);
    }
    return writer;
}

void video_processing(
    const std::string& video_path,
    const std::string& output_id,
    kvald::UNet& model,
    const nlohmann::json& config
)
{
    // Load configuration
    std::vector<int> smoothing_ksize = config["smoothing"]["kernel_size"].get<std::vector<int>>();
    double smoothing_sigma = config["smoothing"]["sigma"];
    std::string device_str = config["train"]["device"];
    torch::Device device(device_str);

    model->to(device);
    model->eval();

    // Get video properties for output
    cv::VideoCapture cap(video_path);
    if (!cap.isOpened())
    {
        throw VideoCaptureError("Cannot open video file: " + video_path);
    }
    int width = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
    int height = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
    double fps = cap.get(cv::CAP_PROP_FPS);
    cap.release();

    // Initialize video writers
    cv::VideoWriter mask_writer = create_video_writer(output_id + "_mask.mp4", fps, width, height, false);
    cv::VideoWriter final_writer = create_video_writer(output_id + "_final.mp4", fps, width, height, true);

    // Initialize Kalman Filter
    kvald::MaskKalmanFilter kf(
        Eigen::Array2i(height, width),
        config["kalman_filter"]["Q"],
        config["kalman_filter"]["R"]
    );

    // Frame queue and threading events
    std::queue<cv::Mat> frame_queue;
    std::mutex queue_mutex;
    std::condition_variable queue_cond;
    std::atomic<bool> stop_reading(false);

    // Start the frame reader thread
    std::thread reader_thread(frame_reader, video_path, std::ref(frame_queue), std::ref(queue_mutex), std::ref(queue_cond), std::ref(stop_reading));

    int frame_count = 0;
    try
    {
        while (true)
        {
            cv::Mat frame;
            {
                std::unique_lock<std::mutex> lock(queue_mutex);
                queue_cond.wait(lock, [&]{ return !frame_queue.empty(); });
                frame = frame_queue.front();
                frame_queue.pop();
            }

            if (frame.empty()) // Sentinel for end of stream or error
            {
                if (stop_reading && frame_queue.empty()) // If reading stopped due to error and queue is empty
                {
                    break;
                }
                else if (!stop_reading) // Normal end of video
                {
                    break;
                }
                else // Error during reading, but queue might still have frames
                {
                    // This case should ideally be handled by throwing from frame_reader
                    // but as a fallback, we break if we get an empty frame and stop_reading is true
                    break;
                }
            }

            frame_count++;
            cv::Mat gray_frame;
            cv::cvtColor(frame, gray_frame, cv::COLOR_BGR2GRAY);
            gray_frame.convertTo(gray_frame, CV_32F, 1.0 / 255.0);

            cv::Mat mask = generate_mask(model, gray_frame, device);
            mask = smooth_output(mask, smoothing_ksize, smoothing_sigma);
            
            // Convert cv::Mat to Eigen::ArrayXf for Kalman Filter
            Eigen::Map<Eigen::MatrixXf> eigen_mask_map(reinterpret_cast<float*>(mask.data), mask.rows, mask.cols);
            Eigen::ArrayXf filtered_mask_eigen = kf.update(Eigen::Map<Eigen::ArrayXf>(eigen_mask_map.data(), eigen_mask_map.size()));

            // Convert Eigen::ArrayXf back to cv::Mat
            cv::Mat filtered_mask_cv(mask.rows, mask.cols, CV_32F);
            Eigen::Map<Eigen::ArrayXf>(reinterpret_cast<float*>(filtered_mask_cv.data), filtered_mask_eigen.size()) = filtered_mask_eigen;

            try
            {
                verify_feedback(filtered_mask_cv);
            }
            catch (const MaskVerificationError& e)
            {
                std::cerr << "Warning: Frame " << frame_count << " failed verification: " << e.what() << std::endl;
                continue;
            }

            cv::Mat mask_output;
            filtered_mask_cv.convertTo(mask_output, CV_8U, 255.0);
            
            cv::Mat final_output = finalize_output(frame, filtered_mask_cv);

            mask_writer.write(mask_output);
            final_writer.write(final_output);
        }
    }
    catch (const VideoProcessingError& e)
    {
        std::cerr << "An error occurred during video processing: " << e.what() << std::endl;
    }
    catch (const std::exception& e)
    {
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
    }

    // Clean up
    stop_reading = true;
    queue_cond.notify_one(); // Notify reader to unblock if waiting
    if (reader_thread.joinable())
    {
        reader_thread.join();
    }
    mask_writer.release();
    final_writer.release();

    if (frame_count == 0)
    {
        throw FrameProcessingError("No frames were read from the video.");
    }

    std::cout << "Processing complete. Videos saved as " << output_id << "_mask.mp4 and " << output_id << "_final.mp4" << std::endl;
}

} // namespace kvald
